KEP-4049: Add storage capacity scoring to VolumeBinding plugin

cupnes opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR is the alpha version of [KEP-4049](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md), incorporating its detailed design and adding basic tests.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Part of https://github.com/kubernetes/enhancements/issues/4049

#### Special notes for your reviewer:

None.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The `StorageCapacityScoring` feature gate was added to score nodes by available storage capacity. It's in alpha and disabled by default. The `VolumeCapacityPriority` alpha feature was replaced with this, and the default behavior was changed. The `VolumeCapacityPriority` preferred a node with the least allocatable, but the `StorageCapacityScoring` preferred a node with the maximum allocatable. See [KEP-4049](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md) for details.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning
```

bells17 committed:
Add storage capacity scoring to VolumeBinding plugin

cupnes commented:
/test all

cupnes commented:
/retest

cupnes commented:
@denkensk @jingxu97 
Could you please review this PR?

xing-yang commented:
/assign @jsafrane @msau42 
/assign

xing-yang commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -470,20 +472,41 @@ func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, po
-				Requested: 0,
-				Capacity:  0,
+	if len(podVolumes.StaticBindings) != 0 || !pl.fts.EnableStorageCapacityScoring {
+		// group static biding volumes by storage class
```
biding -> binding

cupnes replied:
Thank you. I fixed it.

xing-yang commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -470,20 +472,41 @@ func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, po
+			classResources[class].Capacity += storageResource.Capacity
+		}
+	} else {
+		// group dynamic biding volumes by storage class
```
biding -> binding

cupnes replied:
I fixed it.

jsafrane commented on pkg/scheduler/framework/plugins/feature/feature.go:
```diff
@@ -27,4 +27,5 @@ type Features struct {
 	EnableInPlacePodVerticalScaling              bool
 	EnableSidecarContainers                      bool
 	EnableSchedulingQueueHint                    bool
+	EnableStorageCapacityScoring                 bool
```
I don't see where is this field set based on `StorageCapacityScoring` feature gate

cupnes replied:
I have set `EnableStorageCapacityScoring` in pkg/scheduler/framework/plugins/registry.go.

jsafrane commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -470,20 +472,41 @@ func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, po
+					Capacity:  0,
+				}
 			}
+			requestedQty := provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
```
No unit test reaches this code. I wanted to check how the output looks like, but it's not possible to get here.

cupnes replied:
(As pointed out in later comments,) the unit test in pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go was set to skip, so the code did not reach the section you mentioned.
When the feature gate StorageCapacityScoring is enabled, I believe the PreScore function should return nil for both static and dynamic provisioning, so I have modified the PreScore function accordingly.

jsafrane commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -470,20 +472,41 @@ func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, po
 			}
+			requestedQty := provision.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+			classResources[class].Requested += requestedQty.Value()
+			classResources[class].Capacity += provision.Capacity.Capacity.Value()
```
Why is the scheduler *summing* node capacities? If I have a Pod that uses 2 50GiB PVCs of the same StorageClass, and a node with 100GiB capacity for that class, I would expect the node to get full (100%). With added the node capacity for each PVC here (???), the node gets 50% full. That seems very odd. Please add an unit test for this case.

(Question is what capacity to use if each `DynamicProvisions` reports a different capacity for the same storage class... Use `max()` or `min()`?

In static provisioning, the PV capacity is considered, not the node capacity. So summing them makes some sense.

cupnes replied:
I have completed addressing this matter.

> Why is the scheduler summing node capacities?

This was a mistake in the implementation. I have fixed it.

> Please add an unit test for this case.

[I have added it.](https://github.com/kubernetes/kubernetes/pull/128184/files#diff-bf8df5283cbffef0aed1bc2d5b1a0575021e4cf0ce27bd1d70b48dfec1343b1dR866-R901)

> (Question is what capacity to use if each `DynamicProvisions` reports a different capacity for the same storage class... Use `max()` or `min()` or what?

In that case, I think we should use the most recent value, so [I have corrected it accordingly.](https://github.com/kubernetes/kubernetes/pull/128184/files#diff-75e4458878912a7aef522a2eebd4928b91d7622445f2da897119450016cc8a49R504)

jsafrane commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding_test.go:
```diff
@@ -730,6 +747,123 @@ func TestVolumeBinding(t *testing.T) {
+				nil,
+				nil,
+			},
+			wantPreScoreStatus: framework.NewStatus(framework.Skip),
```
with `framework.Skip`, the unit test won't call `Score()` function at all. You need to update `VolumeBinding.PreScore` function to return `nil` instead of `Skip` when there are dynamic volumes to score.

cupnes replied:
As mentioned in the previous comment.

cupnes commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -442,8 +453,14 @@ func (b *volumeBinder) AssumePodVolumes(logger klog.Logger, assumedPod *v1.Pod,
 func (b *volumeBinder) RevertAssumedPodVolumes(podVolumes *PodVolumes) {
+	pvcs := []*v1.PersistentVolumeClaim{}
+	for _, dynamicProvision := range podVolumes.DynamicProvisions {
+		if dynamicProvision != nil {
```
Thank you for your comments. As you pointed out, `dynamicProvision == nil` can never occur. Therefore, the check `if dynamicProvision != nil` is unnecessary and has been removed.

cupnes commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -460,7 +477,12 @@ func (b *volumeBinder) BindPodVolumes(ctx context.Context, assumedPod *v1.Pod, p
+		if dynamicProvision != nil {
+			claimsToProvision = append(claimsToProvision, dynamicProvision.PVC)
+		}
+	}
```
I have addressed it as you pointed out.

cupnes commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -103,13 +103,18 @@ func (b *BindingInfo) StorageResource() *StorageResource {
 
+type DynamicProvision struct {
+	PVC      *v1.PersistentVolumeClaim
+	Capacity *storagev1.CSIStorageCapacity
```
I addressed it.

cupnes commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -942,11 +966,11 @@ func (b *volumeBinder) revertAssumedPVCs(claims []*v1.PersistentVolumeClaim) {
 
 // hasEnoughCapacity checks whether the provisioner has enough capacity left for a new volume of the given size
 // that is available from the node.
```
I addressed it.

cupnes commented:
/retest

cupnes commented:
@xing-yang @jsafrane 
I have finished addressing the points you mentioned, so please review it again.

Additionally, I made the following changes:

- I [added tests](https://github.com/kubernetes/kubernetes/pull/128184/files#diff-bf8df5283cbffef0aed1bc2d5b1a0575021e4cf0ce27bd1d70b48dfec1343b1dR902-R998) for [this part](https://github.com/kubernetes/enhancements/blob/aacf0507d23ba5ea6748e82bebb52f27a1dd8615/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md?plain=1#L445-L448) of the KEP.
- As stated in [this part](https://github.com/kubernetes/enhancements/blob/aacf0507d23ba5ea6748e82bebb52f27a1dd8615/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md?plain=1#L450) of the KEP, when StorageCapacityScoring is enabled, [the default setting is to prefer a node with the maximum allocatable](https://github.com/kubernetes/kubernetes/pull/128184/files#diff-b47c580a8a84d22a66884f23b31d794cba974e89ebc212bd5960b24b3646edf0R197-R206).

xing-yang commented:
@cupnes Can you take a look at this slack discussion?  https://kubernetes.slack.com/archives/C09QZFCE5/p1738363418698329?thread_ts=1737426940.201959&cid=C09QZFCE5

cupnes commented:
I will add a commit about [this comment](https://github.com/kubernetes/kubernetes/pull/128184#issuecomment-2628523423). So, I return this PR to draft.

cupnes commented:
I have corrected the sections that mentioned the alpha version as "1.32" to "1.33".

cupnes committed:
Switch control from VolumeCapacityPriority to StorageCapacityScoring

The tests and comments have also been updated because while
VolumeCapacityPriority preferred a node with the least allocatable,
StorageCapacityScoring preferred a node with the maximum allocatable.

cupnes committed:
Remove VolumeCapacityPriority feature gate

cupnes commented:
/test all

cupnes committed:
Remove the not-even alpha feature setting

Remove the setting for the feature gate that are not even alpha
because Compatibility Versions (KEP-4330) don't allow it.

cupnes commented:
/retest

cupnes commented:
/retest

cupnes commented:
/retest

cupnes commented:
/retest

cupnes commented:
@xing-yang @jsafrane @msau42 
I addressed the deprecation of VolumeCapacityPriority ([Ref](https://kubernetes.slack.com/archives/C09QZFCE5/p1738363418698329?thread_ts=1737426940.201959&cid=C09QZFCE5)). Please review.

wendy-ha18 commented:
Hello @cupnes @xing-yang 
Thanks for actively working on this PR.

Friendly reminder that code freeze is starting at 02:00 UTC Friday 21st March 2025 (about 2 weeks from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time.

Please make sure the PR has both `lgtm` and `approved` labels before the code freeze. Thanks!

cupnes commented on pkg/scheduler/apis/config/v1/defaults.go:
```diff
@@ -192,15 +192,15 @@ func SetDefaults_VolumeBindingArgs(obj *configv1.VolumeBindingArgs) {
 			{
 				Utilization: 0,
-				Score:       0,
+				Score:       int32(config.MaxCustomPriorityScore),
```
I understood. In the alpha version, I will retain the default behavior of VolumeCapacityPriority. However, since I mentioned in the KEP that the default behavior of our StorageCapacityScoring is reversed ([Ref](https://github.com/kubernetes/enhancements/blob/66a23dbda468230aad30c6e2d8f2215f7796a0c3/keps/sig-storage/4049-storage-capacity-scoring-of-nodes-for-dynamic-provisioning/README.md?plain=1#L231)), I plan to switch it in the beta version.

cupnes replied:
I addressed it.

msau42 replied:
Ah, I misunderstood what this was doing. In that case, let's change the default while the feature is still in alpha.

cupnes replied:
I addressed it.

cupnes commented on pkg/features/kube_features.go:
```diff
@@ -642,6 +642,13 @@ const (
+	// kep: https://kep.k8s.io/4049
+	//
+	// Enables scoring nodes by available storage capacity with
+	// StorageCapacityScoring feature gate (dynamic provisioning only).
```
No, that was not my intention. It is an oversight in the comments. I will correct it.

cupnes replied:
I addressed it.

msau42 commented:
Please also add a release note that the VolumeCapacityPriority alpha feature is replaced with this, and we are changing the default behavior.

aaron-prindle commented:
/remove-sig api-machinery

cupnes commented:
@msau42 
I have addressed your comments. Please review them again.

cupnes commented:
/retest

jsafrane commented:
For the record, I tested it in a two node cluster + csi-driver-hostpath.

I can see the scheduler now prefers nodes that have more free capacity when sheduling a 40Gi volume
* i-08ae96773df66c4af has 100Gi of free space
* i-0f6504c648ea98a05 has 60Gi of free space

```
I0312 12:47:23.812228      10 schedule_one.go:794] "Plugin scored node for pod" pod="default/my-csi-app2" plugin="VolumeBinding" node="i-08ae96773df66c4af" score=60
I0312 12:47:23.812266      10 schedule_one.go:794] "Plugin scored node for pod" pod="default/my-csi-app2" plugin="VolumeBinding" node="i-0f6504c648ea98a05" score=34
```

Note to my future self how to test it:

<details>

* Install the driver via [deploy/kubernetes-distributed/deploy.sh](https://github.com/kubernetes-csi/csi-driver-host-path/blob/master/deploy/kubernetes-distributed/deploy.sh)
* Edit the CSIDriver + DaemonSet to enable the capacity tracking:
    `kubectl patch daemonset csi-hostpathplugin -p '{"spec":{"template":{"spec":{"$setElementOrder/containers":[{"name":"csi-provisioner"},{"name":"node-driver-registrar"},{"name":"hostpath"},{"name":"liveness-probe"}],"containers":[{"args":["-v=5","--csi-address=/csi/csi.sock","--feature-gates=Topology=true","--capacity-ownerref-level=0","--node-deployment=true","--strict-topology=true","--immediate-topology=false","--worker-threads=5","--enable-capacity"],"name":"csi-provisioner"}]}}}}' ; kubectl patch csidriver hostpath.csi.k8s.io -p '{"spec":{"storageCapacity":true}}'`
* Run scheduler with `-v 10` to get `Plugin scored node for pod` logs
</details>

msau42 commented:
/kind deprecation
For the VolumeCapacityPriority alpha feature

msau42 commented:
/approve
for config API changes

cupnes commented:
Following [the k8s-ci-robot's comment](https://github.com/kubernetes/kubernetes/pull/128184#issuecomment-2718539575), I assigned.
/assign @alculquicondor

macsko commented:
/assign

macsko commented on pkg/scheduler/framework/plugins/volumebinding/volume_binding.go:
```diff
@@ -470,20 +472,42 @@ func (pl *VolumeBinding) Score(ctx context.Context, cs *framework.CycleState, po
+				continue
+			}
+			class := *provision.PVC.Spec.StorageClassName
+			if _, ok := classResources[class]; !ok {
+				classResources[class] = &StorageResource{
+					Requested: 0,
+					Capacity:  provision.Capacity.Capacity.Value(),
+				}
+			} else {
+				classResources[class].Capacity = provision.Capacity.Capacity.Value()
 			}
```
To make it consistent with StaticBindings code:
```suggestion
			if _, ok := classResources[class]; !ok {
				classResources[class] = &StorageResource{
					Requested: 0,
					Capacity:  0,
				}
			}
			classResources[class].Capacity = provision.Capacity.Capacity.Value()
```

macsko replied:
Btw, shouldn't it be +=?

cupnes replied:
Yes, that's correct. I have addressed it.

cupnes replied:
> Btw, shouldn't it be +=?

No, actually that's not correct. For example, if a Pod requests two 50GB volumes from a StorageClass with 100GB of capacity on a node, this part of the code will be executed twice. In that case, using `+=` would incorrectly set `classResources[class].Capacity` to 200GB.

macsko replied:
Okay, so please add a comment there. And the DynamicProvision.Capacity could be renamed to the NodeCapacity

cupnes replied:
I have addressed it.

aaron-prindle commented:
/remove-sig api-machinery

cupnes commented on pkg/scheduler/framework/plugins/volumebinding/binder.go:
```diff
@@ -377,6 +384,16 @@ func (b *volumeBinder) FindPodVolumes(logger klog.Logger, pod *v1.Pod, podVolume
+// ConvertDynamicProvisionsToPVCs converts a slice of *DynamicProvision to a
+// slice of PersistentVolumeClaim
+func convertDynamicProvisionsToPVCs(dynamicProvisions []*DynamicProvision) []*v1.PersistentVolumeClaim {
+	pvcs := []*v1.PersistentVolumeClaim{}
```
I addressed it.

wendy-ha18 commented:
Hello @cupnes @xing-yang @macsko 

Appreciate all of your efforts with this PR!

A gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure this PR has both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

cupnes commented:
@macsko 
If there are no other problems, could you please give your approval and lgtm?

macsko commented:
/approve
For scheduler changes

macsko commented:
/lgtm

Jefftree commented:
/remove-sig api-machinery
